### PR TITLE
various context related cleanups to rest.Request

### LINKED ISSUE #86590: rest/request.go Request.request retry logic bug?
When retry, r.ctx should be reset to original value.
```
......
for {
		url := r.URL().String()
		req, err := http.NewRequest(r.verb, url, r.body)
		if err != nil {
			return err
		}
                 // when retry, r.ctx should be reset to original value
		if r.timeout > 0 {
			if r.ctx == nil {
				r.ctx = context.Background()
			}
			var cancelFn context.CancelFunc
			r.ctx, cancelFn = context.WithTimeout(r.ctx, r.timeout)
			defer cancelFn()
		}
		if r.ctx != nil {
			req = req.WithContext(r.ctx)
		}
		req.Header = r.headers
                ......
```
Sample code to verify this bug:
```go
// test.go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	var ctx context.Context
	var cancel context.CancelFunc
	for i := 10; i > 0; i-- {
		if ctx == nil {
			ctx = context.Background()
		}
		// ctx = context.Background() // should reset ctx
		ctx, cancel = context.WithTimeout(ctx, time.Second)
		defer cancel()

		// do http request
		select {
		case <-ctx.Done():
			fmt.Println(ctx.Err())
		case <-time.After(time.Duration(i) * 200 * time.Millisecond): // simulate http request time cost
			// expect print OK!!! when i < 5
			fmt.Println("OK!!!")
		}
	}
}
```

go run test.go
```
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
context deadline exceeded
```

### PR DESCRIPTION
**Pull request by mikedanese**
Pre-cleanup pulled out of https://github.com/kubernetes/kubernetes/pull/87597

* Move all usage of r.ctx to the beginning of Do, DoRaw, Stream, Watch
* Move tryThrottle from Do and DoRaw into request()
* Make request() and tryThrottle take a context
* In request(), remove the timeout context setting out of the loop

These changes should be entirely behavior preserving.

/kind cleanup
/sig api-machinery

Fixes #86590

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -660,11 +659,16 @@ func updateURLMetrics(req *Request, resp *http.Response, err error) {
 // Any non-2xx http status code causes an error.  If we get a non-2xx code, we try to convert the body into an APIStatus object.
 // If we can, we return that as an error.  Otherwise, we create an error that lists the http status and the content of the response.
 func (r *Request) Stream() (io.ReadCloser, error) {
+	ctx := context.Background()
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go]: nit, flip this to:
```go
ctx := r.ctx
if ctx == nil {
  ctx = context.Background()
}
```

avoids an allocation, and later when ctx becomes a parameter to this method, we can just drop the `ctx := r.ctx` line

Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/request.go]: oh, unless we require ctx to be non-nil in all invocations... in which case this is temporary/fine

Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/request.go]: Yep. This should be temporary.

**Pull request merged**